core/server/common/src/main/java/alluxio/master/journal/JournalCheckpointThread.java

```diff
@@ -0,0 +1,217 @@
+   */
+  public JournalCheckpointThread(Master master, Journal journal) {
+    mMaster = Preconditions.checkNotNull(master);
+    mJournal = Preconditions.checkNotNull(journal);
+    mShutdownQuietWaitTimeMs =
+        Configuration.getInt(PropertyKey.MASTER_JOURNAL_TAILER_SHUTDOWN_QUIET_WAIT_TIME_MS);
+    mJournalCheckpointSleepTimeMs =
+        Configuration.getInt(PropertyKey.MASTER_JOURNAL_TAILER_SLEEP_TIME_MS);
+    mJournalReader = mJournal.getReader(JournalReaderCreateOptions.defaults().setPrimary(false));
+  }
+
+  /**
+   * Initiates the shutdown of this checkpointer thread, and also waits for it to finish.
+   */
+  public void shutdownAndJoin() {
+    LOG.info("{}: Journal checkpointer shutdown has been initiated.", mMaster.getName());
+    mInitiateShutdown = true;
+
+    try {
+      // Wait for the thread to finish.
+      join();
+    } catch (InterruptedException e) {
+      LOG.error("{}: journal checkpointer shutdown is interrupted.", mMaster.getName(), e);
+      // Kills the master. This can happen in the following two scenarios:
+      // 1. The user Ctrl-C the server.
+      // 2. Zookeeper selects this master as standby before the master finishes the previous
+      //    standby->leader transition. It is safer to crash the server because the behavior is
+      //    undefined to have two journal checkpointer running concurrently.
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * @return the last edit log sequence number read plus 1
+   */
+  public long getNextSequenceNumber() {
+    Preconditions.checkState(mStopped);
+    return mJournalReader.getNextSequenceNumber();
+  }
+
+  @Override
+  public void run() {
+    try {
+      runInternal();
+    } catch (RuntimeException e) {
+      LOG.error("{}: Failed to run journal checkpoint thread, crashing.", mMaster.getName(), e);
+      System.exit(-1);
+    } finally {
+      if (mJournalReader != null) {
+        try {
+          mJournalReader.close();
+        } catch (IOException e) {
+          LOG.warn("{}: Failed to close the journal reader with error {}.", mMaster.getName(),
+              e.getMessage());
+        }
+      }
+    }
+  }
+
+  private void runInternal() {
+    // Keeps reading journal entries. If none is found, sleep for sometime. Periodically write
+    // checkpoints if some conditions are met. The a shutdown signal is receivied, wait until
+    // no new journal entries.
+
+    LOG.info("{}: Journal checkpoint thread started.", mMaster.getName());
+    alluxio.proto.journal.Journal.JournalEntry entry;
+    while (true) {
+      // The start time (ms) for the initiated shutdown.
+      try {
+        entry = mJournalReader.read();
+        if (entry != null) {
+          mMaster.processJournalEntry(entry);
+        }
+      } catch (IOException | InvalidJournalEntryException e) {
+        LOG.warn("{}: Failed to process the journal entry with error {}.", mMaster.getName(),
+            e.getMessage());
```

Comment:
Isn't it possible that the exception is thrown by `read` instead of `processJournalEntry`?